Improve performance of AstNode casting and type checks

gezalore opened the pull request:
Hello,

I spent some time optimising Verilator itself. This is should be the first result of that effort. Purposes of patches are in the commit messages. In short they eliminate all instances of dynamic_cast applied to AstNode via the VN_IS and VN_CAST/VN_CAST_CONST macros.

Git archaeology tells me that the implementation of AstNode casting was similar until 597d28b505ec16f571685347a0050b279ad5781f which suggests it was changed due to some lint checks complaining (cppcheck maybe?), but I feel the 6-13% compilation speed difference I measured is significant enough to warrant the effort/signoff if necessary.

Local testing with --enable-longtests are OK, and Travis non-cron build seem happy.

For completeness, measurements were on an Intel 6440HQ, with fixed CPU frequency so the variance of execution times among 3 runs is under 0.1% as reported by perf stat, so the speed gain is very clear. Host was Ubuntu 17.10 with GCC 7.2.0.

This is my first Verilator contribution so I apologise if I missed something. Please let me know if you would be happy to take similar patches.

Geza

wsnyder commented:
Thanks for your first contribution, and something in the guts too.  Love to see some attention being paid to Verilator performance.

I see your patch has two pieces, I think that's perfect.  I just applied piece number one as it passes for me and are straightforward cleanups.

The change in 597d28b505ec16f571685347a0050b279ad5781f was required to conform with the new C++ standard.  I don't think your change breaks this, but need to validate that and also try it on some older systems, and review more closely.  I see the performance benefit on the second half, so suspect it will all be good.

wsnyder commented:
Please rebase as first half applied.

Old gcc's were happy, I expected them to be unhappy with the templates.  But I'm curious why you need the template & template specializations?  I'd think it would work without them (closer to the old code) since we're defining all of the functions of the appropriate types.  I'm generally of the avoid templates if can school.

Please run using clang and repair, as it's unhappy (./configure CXX=clang++)

     ./V3Ast__gen_impl.h:189:117: error: comparison of two values with different enumeration types ('AstType::en' and 'AstType::bounds') [-Werror,-Wenum-compare]
     template<> inline bool AstNode::privateIs<AstNodeArrayDType>(const AstNode* nodep) { return nodep && (nodep->type() >= AstType::firstNodeArrayDType) && (nodep->type() <= AstType::lastNo...

gezalore commented:
Thank you for the quick feedback. I have fixed the clang complaint and rebased.

You are right looking at just this the templates might seem excessive. The reason I have done it that way is because I had a couple more dependent patches which rely on them being templates. One of them for example parallelises EmitC (which gains another 5% on my compilations against an SSD and many more over a slower file system like NFS). This is achieved by making very minimal changes to  the visitor to remove access to static state and replacing the for loop with a parallel for over a thread pool. In code it reads something like this: 

```
    // Process each module in turn
    v3ThreadPool.foreach<AstNodeModule>(v3Global.rootp()->modulesp(), [](AstNodeModule *nodep){
        if (v3Global.opt.outputSplit()) {
            { EmitCImp fast; fast.main(nodep, false, true); }
            { EmitCImp slow; slow.main(nodep, true, false); }
        } else {
            { EmitCImp both; both.main(nodep, true, true); }
        }
    });
```

I prefer the parallel abstractions be kept templates for better type safety and more use site readability. Also they already rely on standard library templates like std::future. These templates themselves need to do casting underneath and you can't do VN_CAST(foo, T) easily as the cast macro adds the Ast prefix to the type, but accessing the specialised templates fits naturally.

Note: I appreciate the above is a bit experimental and I wanted to put some more thought into it before discussing it in detail (input is welcome though if you have thought). I also appreciate that the above example heavily depends on C++11 being required, but I understand that is coming in the foreseeable future.

wsnyder commented:
Thanks for the updates, will look at these tonight.

We aren't yet accepting requiring use of C++11, with the exception of using --threads in the runtime includes/generated code.  If you want to suggest a push with some of the internal refactoring to e.g. EmitImp to enable this eventually (that still work under pre-C++11) that's fine.

I'm a little surprised that 5% of the time is in V3Emit, but suppose is possible with slow disks.  I wonder if a class that delays all closes to the end would give you most of the same benefit without the risks of threads?  (Probably a tweak and some statics in V3File can do this.)

gezalore commented:
Ok, thanks for the pointers, I will play around with V3Emit a bit more. As a side note: 5% is in V3Emit if you verilate with -Oi to disable inlining (which we do for dev builds to enable incremental compilation of the verilated model via ccache) and -Or (which takes a very long time and has no benefit for our design). With -Oi we have ~8k output files..
